# Capstone Two: Adidas US Sales Forecasting

# 2. Data Wrangling
In this capstone project, we'll develop a sales forecasting model using Adidas 2020-2021 US sales data with the goal to optimize Adidas' sales distribution channels.

In this phase of the project, we'll import, review, clean and prepare the data for the exploratory data analysis in the next phase.

Some of the fundamental questions to resolve in this phase include:

- Do we have all the data needed to tackle the desired question?
- Do we have potentially useful features?
- Are there any fundamental issues with the data that need to be addressed prior to EDA?


## 2.1 Import Modules

In [1]:
# Import pandas, numpy, seaborn and matplotlib.pyplot modules.

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

## 2.2 Load the Data

In [2]:
df = pd.read_excel('Adidas US Sales Datasets.xlsx')
df

Unnamed: 0   Unnamed: 1             Unnamed: 2           Unnamed: 3  \
0            NaN          NaN  Adidas Sales Database                  NaN   
1            NaN          NaN                    NaN                  NaN   
2            NaN          NaN                    NaN                  NaN   
3            NaN     Retailer            Retailer ID         Invoice Date   
4            NaN  Foot Locker                1185732  2020-01-01 00:00:00   
...          ...          ...                    ...                  ...   
9647         NaN  Foot Locker                1185732  2021-01-24 00:00:00   
9648         NaN  Foot Locker                1185732  2021-01-24 00:00:00   
9649         NaN  Foot Locker                1185732  2021-02-22 00:00:00   
9650         NaN  Foot Locker                1185732  2021-02-22 00:00:00   
9651         NaN  Foot Locker                1185732  2021-02-22 00:00:00   

     Unnamed: 4     Unnamed: 5  Unnamed: 6               Unnamed: 7  \
0           NaN            NaN         NaN                      NaN   
1           NaN            NaN         NaN                      NaN   
2           NaN            NaN         NaN                      NaN   
3        Region          State        City                  Product   
4     Northeast       New York    New York    Men's Street Footwear   
...         ...            ...         ...                      ...   
9647  Northeast  New Hampshire  Manchester            Men's Apparel   
9648  Northeast  New Hampshire  Manchester          Women's Apparel   
9649  Northeast  New Hampshire  Manchester    Men's Street Footwear   
9650  Northeast  New Hampshire  Manchester  Men's Athletic Footwear   
9651  Northeast  New Hampshire  Manchester  Women's Street Footwear   

          Unnamed: 8  Unnamed: 9  Unnamed: 10       Unnamed: 11  \
0                NaN         NaN          NaN               NaN   
1                NaN         NaN          NaN               NaN   
2                NaN         NaN          NaN               NaN   
3     Price per Unit  Units Sold  Total Sales  Operating Profit   
4                 50        1200       600000            300000   
...              ...         ...          ...               ...   
9647              50          64         3200             896.0   
9648              41         105         4305            1377.6   
9649              41         184         7544           2791.28   
9650              42          70         2940            1234.8   
9651              29          83         2407            649.89   

           Unnamed: 12   Unnamed: 13  
0                  NaN           NaN  
1                  NaN           NaN  
2                  NaN           NaN  
3     Operating Margin  Sales Method  
4                  0.5      In-store  
...                ...           ...  
9647              0.28        Outlet  
9648              0.32        Outlet  
9649              0.37        Outlet  
9650              0.42        Outlet  
9651              0.27        Outlet  

[9652 rows x 14 columns]

## 2.3 Explore, Clean & Prepare the Data

Based upon the above original dataframe, the 4th row from the top seems to be column names, and the top 3 rows mostly contain missing data with 1 cell containing the database name and therefore could be dropped.  In addition, the first column from the left seems to contain no data and could be dropped.  

In [3]:
# get rid of the top 3 rows.
df = df[3:]

In [4]:
# Check to confirm if the first column contain any information.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9649 entries, 3 to 9651
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   0 non-null      float64
 1   Unnamed: 1   9649 non-null   object 
 2   Unnamed: 2   9649 non-null   object 
 3   Unnamed: 3   9649 non-null   object 
 4   Unnamed: 4   9649 non-null   object 
 5   Unnamed: 5   9649 non-null   object 
 6   Unnamed: 6   9649 non-null   object 
 7   Unnamed: 7   9649 non-null   object 
 8   Unnamed: 8   9649 non-null   object 
 9   Unnamed: 9   9649 non-null   object 
 10  Unnamed: 10  9649 non-null   object 
 11  Unnamed: 11  9649 non-null   object 
 12  Unnamed: 12  9649 non-null   object 
 13  Unnamed: 13  9649 non-null   object 
dtypes: float64(1), object(13)
memory usage: 1.0+ MB


In [5]:
# Since first column doesn't contain any data, let's get rid of it.
df = df.drop(df.columns[0], axis=1)

In [6]:
df

Unnamed: 1   Unnamed: 2           Unnamed: 3 Unnamed: 4     Unnamed: 5  \
3        Retailer  Retailer ID         Invoice Date     Region          State   
4     Foot Locker      1185732  2020-01-01 00:00:00  Northeast       New York   
5     Foot Locker      1185732  2020-01-02 00:00:00  Northeast       New York   
6     Foot Locker      1185732  2020-01-03 00:00:00  Northeast       New York   
7     Foot Locker      1185732  2020-01-04 00:00:00  Northeast       New York   
...           ...          ...                  ...        ...            ...   
9647  Foot Locker      1185732  2021-01-24 00:00:00  Northeast  New Hampshire   
9648  Foot Locker      1185732  2021-01-24 00:00:00  Northeast  New Hampshire   
9649  Foot Locker      1185732  2021-02-22 00:00:00  Northeast  New Hampshire   
9650  Foot Locker      1185732  2021-02-22 00:00:00  Northeast  New Hampshire   
9651  Foot Locker      1185732  2021-02-22 00:00:00  Northeast  New Hampshire   

      Unnamed: 6                 Unnamed: 7      Unnamed: 8  Unnamed: 9  \
3           City                    Product  Price per Unit  Units Sold   
4       New York      Men's Street Footwear              50        1200   
5       New York    Men's Athletic Footwear              50        1000   
6       New York    Women's Street Footwear              40        1000   
7       New York  Women's Athletic Footwear              45         850   
...          ...                        ...             ...         ...   
9647  Manchester              Men's Apparel              50          64   
9648  Manchester            Women's Apparel              41         105   
9649  Manchester      Men's Street Footwear              41         184   
9650  Manchester    Men's Athletic Footwear              42          70   
9651  Manchester    Women's Street Footwear              29          83   

      Unnamed: 10       Unnamed: 11       Unnamed: 12   Unnamed: 13  
3     Total Sales  Operating Profit  Operating Margin  Sales Method  
4          600000            300000               0.5      In-store  
5          500000            150000               0.3      In-store  
6          400000            140000              0.35      In-store  
7          382500            133875              0.35      In-store  
...           ...               ...               ...           ...  
9647         3200             896.0              0.28        Outlet  
9648         4305            1377.6              0.32        Outlet  
9649         7544           2791.28              0.37        Outlet  
9650         2940            1234.8              0.42        Outlet  
9651         2407            649.89              0.27        Outlet  

[9649 rows x 13 columns]

In [7]:
# Reset the index & set the first row as the column name.
df = df.reset_index(drop=True)
df.columns = df.iloc[0]
df.drop(df.index[0], inplace=True)

df.head()

0     Retailer Retailer ID         Invoice Date     Region     State  \
1  Foot Locker     1185732  2020-01-01 00:00:00  Northeast  New York   
2  Foot Locker     1185732  2020-01-02 00:00:00  Northeast  New York   
3  Foot Locker     1185732  2020-01-03 00:00:00  Northeast  New York   
4  Foot Locker     1185732  2020-01-04 00:00:00  Northeast  New York   
5  Foot Locker     1185732  2020-01-05 00:00:00  Northeast  New York   

0      City                    Product Price per Unit Units Sold Total Sales  \
1  New York      Men's Street Footwear             50       1200      600000   
2  New York    Men's Athletic Footwear             50       1000      500000   
3  New York    Women's Street Footwear             40       1000      400000   
4  New York  Women's Athletic Footwear             45        850      382500   
5  New York              Men's Apparel             60        900      540000   

0 Operating Profit Operating Margin Sales Method  
1           300000              0.5     In-store  
2           150000              0.3     In-store  
3           140000             0.35     In-store  
4           133875             0.35     In-store  
5           162000              0.3     In-store

In [8]:
# Reset the index starting from 0.
df = df.reset_index(drop=True)
df.head()

0     Retailer Retailer ID         Invoice Date     Region     State  \
0  Foot Locker     1185732  2020-01-01 00:00:00  Northeast  New York   
1  Foot Locker     1185732  2020-01-02 00:00:00  Northeast  New York   
2  Foot Locker     1185732  2020-01-03 00:00:00  Northeast  New York   
3  Foot Locker     1185732  2020-01-04 00:00:00  Northeast  New York   
4  Foot Locker     1185732  2020-01-05 00:00:00  Northeast  New York   

0      City                    Product Price per Unit Units Sold Total Sales  \
0  New York      Men's Street Footwear             50       1200      600000   
1  New York    Men's Athletic Footwear             50       1000      500000   
2  New York    Women's Street Footwear             40       1000      400000   
3  New York  Women's Athletic Footwear             45        850      382500   
4  New York              Men's Apparel             60        900      540000   

0 Operating Profit Operating Margin Sales Method  
0           300000              0.5     In-store  
1           150000              0.3     In-store  
2           140000             0.35     In-store  
3           133875             0.35     In-store  
4           162000              0.3     In-store

In [9]:
nullcounts = df.isna().sum()
nullcounts

0
Retailer            0
Retailer ID         0
Invoice Date        0
Region              0
State               0
City                0
Product             0
Price per Unit      0
Units Sold          0
Total Sales         0
Operating Profit    0
Operating Margin    0
Sales Method        0
dtype: int64

In [10]:
df.shape

(9648, 13)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9648 entries, 0 to 9647
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Retailer          9648 non-null   object
 1   Retailer ID       9648 non-null   object
 2   Invoice Date      9648 non-null   object
 3   Region            9648 non-null   object
 4   State             9648 non-null   object
 5   City              9648 non-null   object
 6   Product           9648 non-null   object
 7   Price per Unit    9648 non-null   object
 8   Units Sold        9648 non-null   object
 9   Total Sales       9648 non-null   object
 10  Operating Profit  9648 non-null   object
 11  Operating Margin  9648 non-null   object
 12  Sales Method      9648 non-null   object
dtypes: object(13)
memory usage: 980.0+ KB


The database has a total of 13 columns and 9468 rows.  Good news is that there's no missing data.  Bad news is that all the data type is object and that obviously doesn't work for columns like *<span style="color:gray">Invoice Date</span>*, *<span style="color:gray">Price per Unit</span>*, *<span style="color:gray">Units Sold</span>*, *<span style="color:gray">Total Sales</span>*, *<span style="color:gray">Operating Profit</span>*, and *<span style="color:gray">Operating Margin</span>*.

In [12]:
# Change data type for "Invoice Date" to datetime.
df['Invoice Date'] = pd.to_datetime(df['Invoice Date'])

In [13]:
# Change data type for "Price per Unit", "Total Sales", "Operating Profit" & "Operating Margin" columnes to numeric.
data_columns = ['Price per Unit', 'Total Sales', 'Operating Profit', 'Operating Margin']
df[data_columns] = df[data_columns].apply(pd.to_numeric, errors='coerce')

# Change the data type for "Units Sold" to integer.
df['Units Sold'] = df['Units Sold'].astype(int)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9648 entries, 0 to 9647
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Retailer          9648 non-null   object        
 1   Retailer ID       9648 non-null   object        
 2   Invoice Date      9648 non-null   datetime64[ns]
 3   Region            9648 non-null   object        
 4   State             9648 non-null   object        
 5   City              9648 non-null   object        
 6   Product           9648 non-null   object        
 7   Price per Unit    9648 non-null   float64       
 8   Units Sold        9648 non-null   int32         
 9   Total Sales       9648 non-null   float64       
 10  Operating Profit  9648 non-null   float64       
 11  Operating Margin  9648 non-null   float64       
 12  Sales Method      9648 non-null   object        
dtypes: datetime64[ns](1), float64(4), int32(1), object(7)
memory usage: 942.3+ KB


In [15]:
df.describe()

0      Price per Unit   Units Sold    Total Sales  Operating Profit  \
count     9648.000000  9648.000000    9648.000000       9648.000000   
mean        45.216625   256.930037   93273.437500      34425.244761   
std         14.705397   214.252030  141916.016727      54193.113713   
min          7.000000     0.000000       0.000000          0.000000   
25%         35.000000   106.000000    4254.500000       1921.752500   
50%         45.000000   176.000000    9576.000000       4371.420000   
75%         55.000000   350.000000  150000.000000      52062.500000   
max        110.000000  1275.000000  825000.000000     390000.000000   

0      Operating Margin  
count       9648.000000  
mean           0.422991  
std            0.097197  
min            0.100000  
25%            0.350000  
50%            0.410000  
75%            0.490000  
max            0.800000

Based upon the above descriptive data, there doesn't seem to be any obvious outliers.  The only interesting thing worth exploring is that some of the invoices have 0 units sold.  If nothing is sold, then there shouldn't be an invoice, or that particular product shouldn't be included in the invoice.  So let's take a look at all the entries with 0 units sold to see if the details provide a better explanation.

In [16]:
df_0 = df[df['Units Sold']==0]
df_0

0        Retailer Retailer ID Invoice Date   Region     State   City  \
1019  Foot Locker     1185732   2021-06-05  Midwest  Nebraska  Omaha   
1025  Foot Locker     1185732   2021-06-11  Midwest  Nebraska  Omaha   
4907  Foot Locker     1185732   2021-06-05  Midwest  Nebraska  Omaha   
4913  Foot Locker     1185732   2021-06-11  Midwest  Nebraska  Omaha   

0                       Product  Price per Unit  Units Sold  Total Sales  \
1019  Women's Athletic Footwear            35.0           0          0.0   
1025  Women's Athletic Footwear            30.0           0          0.0   
4907  Women's Athletic Footwear            33.0           0          0.0   
4913  Women's Athletic Footwear            27.0           0          0.0   

0     Operating Profit  Operating Margin Sales Method  
1019               0.0              0.40       Outlet  
1025               0.0              0.40       Outlet  
4907               0.0              0.55       Online  
4913               0.0              0.53       Online

All the 4 zero units sold entries are for Foot Locker with Retailer ID of 1185732 in Omaha, Nebraska.  Let's take a look at the all entries for Foot Locker in Omaha dated 2021-06-05 & 2021-06-11 to see if we can make sense of these 0 units sold entries.

In [17]:
df_footlocker = df[(df['Retailer'] == 'Foot Locker') & (df['City'] == 'Omaha') & 
                   ((df['Invoice Date'] == '2021-06-05')|(df['Invoice Date'] == '2021-06-11'))]
df_footlocker

0        Retailer Retailer ID Invoice Date   Region     State   City  \
1019  Foot Locker     1185732   2021-06-05  Midwest  Nebraska  Omaha   
1025  Foot Locker     1185732   2021-06-11  Midwest  Nebraska  Omaha   
4907  Foot Locker     1185732   2021-06-05  Midwest  Nebraska  Omaha   
4913  Foot Locker     1185732   2021-06-11  Midwest  Nebraska  Omaha   

0                       Product  Price per Unit  Units Sold  Total Sales  \
1019  Women's Athletic Footwear            35.0           0          0.0   
1025  Women's Athletic Footwear            30.0           0          0.0   
4907  Women's Athletic Footwear            33.0           0          0.0   
4913  Women's Athletic Footwear            27.0           0          0.0   

0     Operating Profit  Operating Margin Sales Method  
1019               0.0              0.40       Outlet  
1025               0.0              0.40       Outlet  
4907               0.0              0.55       Online  
4913               0.0              0.53       Online

There is nothing else on the 2021-06-05 and 2021-06-11 invoices for Foot Locker Omaha location other than the above 4 zero Units Sold entries.  While we are still not able to make sense why these zero dollar entries are there in the first place, it probably won't cause any damage to leave these zero dollar entries in the file.  Therefore, let's just keep these entries for now.

Next, we want to make sure that all numbers in the file are calculated correctly.  Since <span style="color:blue">Price per Unit x Units Sold = Total Sales</span> and <span style="color:blue">Operating Profit / Total Sales = Operating Margin</span>, we will manually calculate the total sales value and operating margin for each data entry, and then compare the calculated values to the original values in the file to ensure that all numbers are correct.

In [18]:
# Calculate Total Sales 2 using the above stated formula, compare to the original Total Sales value, and print 
# all the rows where the calculated Total Sales 2 value doesn't equal the origianl Total Sales value.
df['Total Sales 2'] = df['Price per Unit'] * df['Units Sold']
df['Total Sales Check'] = df['Total Sales 2'] - df['Total Sales']
df[df['Total Sales Check']!=0]

0        Retailer Retailer ID Invoice Date     Region          State  \
0     Foot Locker     1185732   2020-01-01  Northeast       New York   
1     Foot Locker     1185732   2020-01-02  Northeast       New York   
2     Foot Locker     1185732   2020-01-03  Northeast       New York   
3     Foot Locker     1185732   2020-01-04  Northeast       New York   
4     Foot Locker     1185732   2020-01-05  Northeast       New York   
...           ...         ...          ...        ...            ...   
3883  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3884  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3885  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3886  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3887  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   

0           City                    Product  Price per Unit  Units Sold  \
0       New York      Men's Street Footwear            50.0        1200   
1       New York    Men's Athletic Footwear            50.0        1000   
2       New York    Women's Street Footwear            40.0        1000   
3       New York  Women's Athletic Footwear            45.0         850   
4       New York              Men's Apparel            60.0         900   
...          ...                        ...             ...         ...   
3883  Manchester              Men's Apparel            65.0         200   
3884  Manchester            Women's Apparel            55.0         300   
3885  Manchester      Men's Street Footwear            55.0         575   
3886  Manchester    Men's Athletic Footwear            55.0         225   
3887  Manchester    Women's Street Footwear            45.0         275   

0     Total Sales  Operating Profit  Operating Margin Sales Method  \
0        600000.0          300000.0              0.50     In-store   
1        500000.0          150000.0              0.30     In-store   
2        400000.0          140000.0              0.35     In-store   
3        382500.0          133875.0              0.35     In-store   
4        540000.0          162000.0              0.30     In-store   
...           ...               ...               ...          ...   
3883     130000.0           32500.0              0.25       Outlet   
3884     165000.0           49500.0              0.30       Outlet   
3885     316250.0          110687.5              0.35       Outlet   
3886     123750.0           43312.5              0.35       Outlet   
3887     123750.0           30937.5              0.25       Outlet   

0     Total Sales 2  Total Sales Check  
0           60000.0          -540000.0  
1           50000.0          -450000.0  
2           40000.0          -360000.0  
3           38250.0          -344250.0  
4           54000.0          -486000.0  
...             ...                ...  
3883        13000.0          -117000.0  
3884        16500.0          -148500.0  
3885        31625.0          -284625.0  
3886        12375.0          -111375.0  
3887        12375.0          -111375.0  

[3886 rows x 15 columns]

There are 3886 rows where the calculated *<span style="color:gray">Total Sales 2</span>* value doesn't match the original *<span style="color:gray">Total Sales</span>* value.  Upon a closer look at the top 5 and bottom 5 mismatched rows, it appears that the calculated *<span style="color:gray">Total Sales 2</span>* value is 1/10 of the original *<span style="color:gray">Total Sales</span>* value.  Next, let's verify whether that applies to all 3886 rows.

In [19]:
df_sales_dis = df[df['Total Sales Check']!=0]
df_sales_dis['Discrepancy Scale'] = df_sales_dis['Total Sales'] / df_sales_dis['Total Sales 2']
df_sales_dis

C:\Users\wfgyi\AppData\Local\Temp\ipykernel_13616\196164795.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sales_dis['Discrepancy Scale'] = df_sales_dis['Total Sales'] / df_sales_dis['Total Sales 2']


0        Retailer Retailer ID Invoice Date     Region          State  \
0     Foot Locker     1185732   2020-01-01  Northeast       New York   
1     Foot Locker     1185732   2020-01-02  Northeast       New York   
2     Foot Locker     1185732   2020-01-03  Northeast       New York   
3     Foot Locker     1185732   2020-01-04  Northeast       New York   
4     Foot Locker     1185732   2020-01-05  Northeast       New York   
...           ...         ...          ...        ...            ...   
3883  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3884  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3885  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3886  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3887  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   

0           City                    Product  Price per Unit  Units Sold  \
0       New York      Men's Street Footwear            50.0        1200   
1       New York    Men's Athletic Footwear            50.0        1000   
2       New York    Women's Street Footwear            40.0        1000   
3       New York  Women's Athletic Footwear            45.0         850   
4       New York              Men's Apparel            60.0         900   
...          ...                        ...             ...         ...   
3883  Manchester              Men's Apparel            65.0         200   
3884  Manchester            Women's Apparel            55.0         300   
3885  Manchester      Men's Street Footwear            55.0         575   
3886  Manchester    Men's Athletic Footwear            55.0         225   
3887  Manchester    Women's Street Footwear            45.0         275   

0     Total Sales  Operating Profit  Operating Margin Sales Method  \
0        600000.0          300000.0              0.50     In-store   
1        500000.0          150000.0              0.30     In-store   
2        400000.0          140000.0              0.35     In-store   
3        382500.0          133875.0              0.35     In-store   
4        540000.0          162000.0              0.30     In-store   
...           ...               ...               ...          ...   
3883     130000.0           32500.0              0.25       Outlet   
3884     165000.0           49500.0              0.30       Outlet   
3885     316250.0          110687.5              0.35       Outlet   
3886     123750.0           43312.5              0.35       Outlet   
3887     123750.0           30937.5              0.25       Outlet   

0     Total Sales 2  Total Sales Check  Discrepancy Scale  
0           60000.0          -540000.0               10.0  
1           50000.0          -450000.0               10.0  
2           40000.0          -360000.0               10.0  
3           38250.0          -344250.0               10.0  
4           54000.0          -486000.0               10.0  
...             ...                ...                ...  
3883        13000.0          -117000.0               10.0  
3884        16500.0          -148500.0               10.0  
3885        31625.0          -284625.0               10.0  
3886        12375.0          -111375.0               10.0  
3887        12375.0          -111375.0               10.0  

[3886 rows x 16 columns]

In [20]:
# Verify if all values in the "Discrepany Scale" column equal 10.  
df_sales_dis[round(df_sales_dis['Discrepancy Scale'])==10]

0        Retailer Retailer ID Invoice Date     Region          State  \
0     Foot Locker     1185732   2020-01-01  Northeast       New York   
1     Foot Locker     1185732   2020-01-02  Northeast       New York   
2     Foot Locker     1185732   2020-01-03  Northeast       New York   
3     Foot Locker     1185732   2020-01-04  Northeast       New York   
4     Foot Locker     1185732   2020-01-05  Northeast       New York   
...           ...         ...          ...        ...            ...   
3883  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3884  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
3885  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3886  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
3887  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   

0           City                    Product  Price per Unit  Units Sold  \
0       New York      Men's Street Footwear            50.0        1200   
1       New York    Men's Athletic Footwear            50.0        1000   
2       New York    Women's Street Footwear            40.0        1000   
3       New York  Women's Athletic Footwear            45.0         850   
4       New York              Men's Apparel            60.0         900   
...          ...                        ...             ...         ...   
3883  Manchester              Men's Apparel            65.0         200   
3884  Manchester            Women's Apparel            55.0         300   
3885  Manchester      Men's Street Footwear            55.0         575   
3886  Manchester    Men's Athletic Footwear            55.0         225   
3887  Manchester    Women's Street Footwear            45.0         275   

0     Total Sales  Operating Profit  Operating Margin Sales Method  \
0        600000.0          300000.0              0.50     In-store   
1        500000.0          150000.0              0.30     In-store   
2        400000.0          140000.0              0.35     In-store   
3        382500.0          133875.0              0.35     In-store   
4        540000.0          162000.0              0.30     In-store   
...           ...               ...               ...          ...   
3883     130000.0           32500.0              0.25       Outlet   
3884     165000.0           49500.0              0.30       Outlet   
3885     316250.0          110687.5              0.35       Outlet   
3886     123750.0           43312.5              0.35       Outlet   
3887     123750.0           30937.5              0.25       Outlet   

0     Total Sales 2  Total Sales Check  Discrepancy Scale  
0           60000.0          -540000.0               10.0  
1           50000.0          -450000.0               10.0  
2           40000.0          -360000.0               10.0  
3           38250.0          -344250.0               10.0  
4           54000.0          -486000.0               10.0  
...             ...                ...                ...  
3883        13000.0          -117000.0               10.0  
3884        16500.0          -148500.0               10.0  
3885        31625.0          -284625.0               10.0  
3886        12375.0          -111375.0               10.0  
3887        12375.0          -111375.0               10.0  

[3886 rows x 16 columns]

Here are our observations based upon the above calculation and data.

1. There are a total of 3886 rows where the calculated *<span style="color:gray">Total Sales 2</span>* value doesn't match the original *<span style="color:gray">Total Sales</span>* value.  
2. All discrepancies are caused by the fact that the calculated *<span style="color:gray">Total Sales 2</span>* value is 1/10 of the original *<span style="color:gray">Total Sales</span>* value.
3. Another interesting point is that all these 3886 rows are on the top of the dataframe from row # 0 to row # 3887, with the exception of rows # 1019 & 1025 with zero Units Sold.  

In a real business environment, we'll need to pull additional sales data and/or order details to conduct further analysis.  It could be the *<span style="color:gray">Unit Price</span>*, or *<span style="color:gray">Units Sold</span>*, or *<span style="color:gray">Total Sales</span>* values that were entered incorrectly.  However, without additional data, we can't simply conclude one column vs. the other two columns were incorrect and take actions accordingly.  Any unfounded actions taken now could negatively impact the forecasting result or segmentation analysis we'll perform in the following phases of the project.  Therefore, in order to proceed with the project, we have to make the assumption that all three columns of original data are entered correctly, and the discrepancy between the calculated *<span style="color:gray">Total Sales 2</span>*  value and the original *<span style="color:gray">Total Sales</span>* value is due to situations beyond my understanding based upon the limited information we currently possess.

Next, let's calculate the operating margin and see if the calculated values match the original values in the dataframe.

In [21]:
# Delete the added "Total Sales 2", "Total Sales Check" & "Discrepancy Scale" columns.
df.drop(["Total Sales 2", "Total Sales Check"], axis=1, inplace=True)

In [22]:
# Calculate Operating Margin 2 using the above stated formula, compare to the original Operating Margin value, and print 
# all the rows where the calculated Operating Margin 2 value doesn't equal the origianl Operating Margin value.
df['Operating Margin 2'] = df['Operating Profit'] / df['Total Sales']
df['Operating Margin Check'] = df['Operating Margin 2'] - df['Operating Margin']
df[df['Operating Margin Check']!=0]

0          Retailer Retailer ID Invoice Date     Region          State  \
142       West Gear     1128299   2020-11-03      South          Texas   
152       West Gear     1128299   2020-11-13       West     California   
154       West Gear     1128299   2020-11-15       West     California   
160       West Gear     1128299   2020-12-27       West     California   
392   Sports Direct     1197831   2021-09-16      South          Texas   
...             ...         ...          ...        ...            ...   
9598         Amazon     1185732   2021-06-19  Northeast  New Hampshire   
9603         Amazon     1185732   2021-07-17  Northeast  New Hampshire   
9609         Amazon     1185732   2021-08-18  Northeast  New Hampshire   
9616         Amazon     1185732   2021-09-17  Northeast  New Hampshire   
9633    Foot Locker     1185732   2021-12-17  Northeast  New Hampshire   

0              City                  Product  Price per Unit  Units Sold  \
142         Houston  Women's Street Footwear            50.0         775   
152   San Francisco    Men's Street Footwear            50.0         850   
154   San Francisco  Women's Street Footwear            60.0         700   
160   San Francisco  Women's Street Footwear            65.0         725   
392          Dallas    Men's Street Footwear            45.0         825   
...             ...                      ...             ...         ...   
9598     Manchester  Men's Athletic Footwear            40.0         140   
9603     Manchester    Men's Street Footwear            49.0         224   
9609     Manchester    Men's Street Footwear            48.0         203   
9616     Manchester  Men's Athletic Footwear            43.0         116   
9633     Manchester    Men's Street Footwear            49.0         200   

0     Total Sales  Operating Profit  Operating Margin Sales Method  \
142      387500.0         135625.00              0.35       Outlet   
152      425000.0         148750.00              0.35       Outlet   
154      420000.0         147000.00              0.35       Outlet   
160      471250.0         164937.50              0.35       Outlet   
392      371250.0         148500.00              0.40       Outlet   
...           ...               ...               ...          ...   
9598       5600.0           2632.00              0.47       Outlet   
9603      10976.0           4609.92              0.42       Outlet   
9609       9744.0           4482.24              0.46       Outlet   
9616       4988.0           2094.96              0.42       Outlet   
9633       9800.0           4606.00              0.47       Outlet   

0     Operating Margin 2  Operating Margin Check  
142                 0.35           -5.551115e-17  
152                 0.35           -5.551115e-17  
154                 0.35           -5.551115e-17  
160                 0.35            5.551115e-17  
392                 0.40            5.551115e-17  
...                  ...                     ...  
9598                0.47           -5.551115e-17  
9603                0.42           -5.551115e-17  
9609                0.46           -5.551115e-17  
9616                0.42           -5.551115e-17  
9633                0.47           -5.551115e-17  

[848 rows x 15 columns]

In [23]:
# Check out the discrepancies between the original Operating Margin values and calculated Operating Margin 2 values.  
df[df['Operating Margin Check']!=0]['Operating Margin Check'].describe()

count    8.440000e+02
mean     3.124147e-19
std      5.754918e-17
min     -1.110223e-16
25%     -5.551115e-17
50%      2.775558e-17
75%      5.551115e-17
max      1.110223e-16
Name: Operating Margin Check, dtype: float64

There are 848 rows where the original *<span style="color:gray">Operating Margin</span>* value doesn't match the calculated *<span style="color:gray">Operating Margin 2</span>* value.  Upon a closer look, the difference is extremely small, which might have been caused by rounding issues and decimal differences.  Therefore, we will ignore the tiny discrepancies in the operating margin values and use the original *<span style="color:gray">Operating Margin</span>* values as is.

In [24]:
# Delete the added "Operating Margin 2", "Operating Margin Check" columns.
df.drop(["Operating Margin 2", "Operating Margin Check"], axis=1, inplace=True)
df

0        Retailer Retailer ID Invoice Date     Region          State  \
0     Foot Locker     1185732   2020-01-01  Northeast       New York   
1     Foot Locker     1185732   2020-01-02  Northeast       New York   
2     Foot Locker     1185732   2020-01-03  Northeast       New York   
3     Foot Locker     1185732   2020-01-04  Northeast       New York   
4     Foot Locker     1185732   2020-01-05  Northeast       New York   
...           ...         ...          ...        ...            ...   
9643  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
9644  Foot Locker     1185732   2021-01-24  Northeast  New Hampshire   
9645  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
9646  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   
9647  Foot Locker     1185732   2021-02-22  Northeast  New Hampshire   

0           City                    Product  Price per Unit  Units Sold  \
0       New York      Men's Street Footwear            50.0        1200   
1       New York    Men's Athletic Footwear            50.0        1000   
2       New York    Women's Street Footwear            40.0        1000   
3       New York  Women's Athletic Footwear            45.0         850   
4       New York              Men's Apparel            60.0         900   
...          ...                        ...             ...         ...   
9643  Manchester              Men's Apparel            50.0          64   
9644  Manchester            Women's Apparel            41.0         105   
9645  Manchester      Men's Street Footwear            41.0         184   
9646  Manchester    Men's Athletic Footwear            42.0          70   
9647  Manchester    Women's Street Footwear            29.0          83   

0     Total Sales  Operating Profit  Operating Margin Sales Method  
0        600000.0         300000.00              0.50     In-store  
1        500000.0         150000.00              0.30     In-store  
2        400000.0         140000.00              0.35     In-store  
3        382500.0         133875.00              0.35     In-store  
4        540000.0         162000.00              0.30     In-store  
...           ...               ...               ...          ...  
9643       3200.0            896.00              0.28       Outlet  
9644       4305.0           1377.60              0.32       Outlet  
9645       7544.0           2791.28              0.37       Outlet  
9646       2940.0           1234.80              0.42       Outlet  
9647       2407.0            649.89              0.27       Outlet  

[9648 rows x 13 columns]

At this point of time, our data has been cleaned, organized, verified and prepared to be worked on in coming phases of the project.  Let's export it.

In [25]:
df.to_csv("df.csv")